### PR TITLE
Copter: Prevent position controller madness caused by bad MAVLink input

### PR DESCRIPTION
**Pull request by peterbarker**
This PR prevents a *specific* problem which caused a user's vehicle to fly away.

The user input a bad value to the vehicle in GUIDED mode, which caused the position controller state to become corrupt - bunches of NaNs logged in the dataflash log etc.

The tests included in this PR recreate that user's input, but does not look at other supplied fields - accel, yaw and yaw-rate, all of which could create the same problem.

I left it here because there are other ways to accomplish the same thing - most notably, dropping it on the set_pos_vel_acc methods on guided mode (and making them boolean if they're not already) would also stop bad things getting in from scripting.  IMO we should also add periodic validation of the position controller's state - forcing an internal error and reset if we detect things have gone pear-shaped; this is a belts-and-braces approach because bugs happen.

image

image

image

### CODE REVIEW

**Comment by peterbarker** [comment]
Log is available here: https://discord.com/channels/674039678562861068/1110563353027412029

File: ArduCopter/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -1258,7 +1273,13 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
         Vector3f vel_vector;
         if (!vel_ignore) {
             // convert to cm
```
Comment (rmackay9) [inline ArduCopter/GCS_Mavlink.cpp:1275]: need to fix this comment maybe?

File: ArduCopter/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -1097,6 +1097,21 @@ void GCS_MAVLINK_Copter::handle_manual_control_axes(const mavlink_manual_control
     manual_override(copter.channel_yaw, packet.r, 1000, 2000, tnow);
 }
 
+// this method is to guard against completely ridiculous values being
```
Comment (rmackay9) [inline ArduCopter/GCS_Mavlink.cpp:1100]: I think we should change this comment to something like, "sanity check velocity or acceleration vector components are all number (e.g. not NaN) and below 1000.  vec argument units are in meters."

File: ArduCopter/GCS_Mavlink.h [hunk 3]
```diff
@@ -55,6 +55,9 @@ class GCS_MAVLINK_Copter : public GCS_MAVLINK
 
 private:
 
+    // returns false if the input vector (in m/s/s or m/s/s) looks invalid
```
Comment (rmackay9) [inline ArduCopter/GCS_Mavlink.h:58]: let's put the same comment in the .h and .cpp.

**Comment by rmackay9** [comment]
Thanks very much for finding and fixing this.

As you say we should consider whether we want this protection at this level (which protects the controllers when the inputs are coming in via MAVLink messages) or whether it should be lower down which could provide protection regardless of the source of the targets.

FYI @lthall

File: ArduCopter/GCS_Mavlink.cpp (hunk 2 above)
Comment (peterbarker) [reply ArduCopter/GCS_Mavlink.cpp]: Done - but I've used the correct units

File: ArduCopter/GCS_Mavlink.h (hunk 3 above)
Comment (peterbarker) [reply ArduCopter/GCS_Mavlink.h]: Done

**Review by rmackay9 — APPROVED** [review]

**Comment by rmackay9** [comment]
FYI @lthall I've merged this.. hope that's OK.

**Comment by lthall** [comment]
yep, no problems

**Pull request merged**